### PR TITLE
[SC-304] SC custom resource to manage products

### PR DESCRIPTION
**Pull request by zaro0508**
AWS cloudformation does not support product deprecation so we've been doing
it manually however that's been an error prone process. We need an
automated way to deprecate products versions on every update.

This PR provides a general custom resource to manage service catalog 
products.  A function has been added to deprecate product versions.

### CODE REVIEW

File: product_provider/app.py [hunk 1]
```diff
@@ -1,42 +1,183 @@
...
 
-        Context doc: https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+  return properties
 
-    Returns
-    ------
-    API Gateway Lambda Proxy Output Format: dict
+def get_latest_provisioning_artifact(provisioning_artifacts):
+  '''
+  Get the latest the latest version of a product's provisioning artifact.
```
Comment (tthyer) [inline product_provider/app.py]: typo

File: product_provider/app.py [hunk 2]
```diff
@@ -1,42 +1,183 @@
...
+
+  :param product_info: response from https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/servicecatalog.html#ServiceCatalog.Client.describe_product_as_admin
+  :return: A list of the product's provisioned artifacts
+  '''
+  ProductId = product_info['ProductViewDetail']['ProductViewSummary']['ProductId']
+  summaries = product_info['ProvisioningArtifactSummaries']
+  artifacts = []
+  for summary in summaries:
+    # remove `v` from the name to get semver
+    semantic_version = summary['Name'][1:]
```
Comment (tthyer) [inline product_provider/app.py:105]: May want to check whether the string starts with v before assuming `[1:]`

File: product_provider/app.py [hunk 3]
```diff
@@ -1,42 +1,183 @@
...
+      "ProvisioningArtifactId": summary["Id"],
+      "Name": semantic_version,
+      "Description": summary["Description"]
     }
+    artifacts.append(artifact)
+
+  return artifacts
+
+
+def get_artifacts_to_update(provisioning_artifacts, action):
```
Comment (tthyer) [inline product_provider/app.py]: should action have a default value?

File: product_provider/app.py [hunk 4]
```diff
@@ -1,42 +1,183 @@
...
+      ProvisioningArtifactId=artifact['ProvisioningArtifactId'],
+      Active=active,
+      Guidance=guidance
+    )
+
+@helper.create
+@helper.update
+def create_or_update(event, context):
+  '''Handles customm resource create and update events'''
+  log.debug('Received event: ' + json.dumps(event, sort_keys=False))
```
Comment (tthyer) [inline product_provider/app.py]: can be an f-string too 😁

**Review by tthyer — APPROVED** [review]
This looks good. Might want to share this repository with our contact at AWS -- this functionality should be part of how SC works.

File: product_provider/app.py (hunk 1 above)
Comment (zaro0508) [reply product_provider/app.py]: fixed

File: product_provider/app.py (hunk 3 above)
Comment (zaro0508) [reply product_provider/app.py]: done

File: product_provider/app.py (hunk 4 above)
Comment (zaro0508) [reply product_provider/app.py]: fixed

File: product_provider/app.py (hunk 2 above)
Comment (zaro0508) [reply product_provider/app.py:105]: it doesn't matter what the name is, we check that it's a valid semantic version on the next line.

**Pull request merged**